### PR TITLE
Allow multiple IPs in Consul Discovery Service

### LINKED ISSUE #576: Help incorrectly shows ability to specify multiple consul IPs
The help message for Swarm shows that you can specify multiple consul IPs (which would be nice): https://github.com/docker/swarm/blob/master/help.go#L21

However, it doesn't actually (yet): 

`swarm version 0.2.0 (HEAD)`

```
$ /usr/local/bin/swarm manage -H tcp://0.0.0.0:2375 --strategy spread consul://10.0.0.5,10.0.1.5/swarm
INFO[0000] Listening for HTTP                            addr=0.0.0.0:2375 proto=tcp
FATA[0000] Put http://10.0.0.5,10.0.1.5/v1/kv/swarm/: dial tcp: lookup 10.0.0.5,10.0.1.5: no such host
```

Trying the command with each IP on it's own works.

Not sure if it was put there in anticipation of the feature, but it's a bit confusing that it's there but doesn't work.

I saw this PR: https://github.com/docker/swarm/pull/386. Not sure if it is being implemented from someone else.

### PR DESCRIPTION
**Pull request by abronan**
Fixes #576 

Open to discussion, because this is mostly hacking our way around the Go consul client.. Needs some testing as well.

### CODE REVIEW

**Comment by abronan** [comment]
Needs some gofmt..

File: discovery/consul/consul.go [hunk 1]
```diff
@@ -103,8 +166,12 @@ func (s *ConsulDiscoveryService) waitForChange() <-chan uint64 {
...
 			_, meta, err := kv.List(s.prefix, option)
 			if err != nil {
-				log.WithField("name", "consul").Errorf("Discovery error: %v", err)
-				break
+				err = s.switchConsulAgent()
+				if err != nil {
+					log.WithField("name", "consul").Errorf("Discovery error: %v", err)
+					break
+				}
+				continue RETRY
```
Comment (vieux) [inline discovery/consul/consul.go]: I don't see a `for` within a `for` is this label necessary ?

File: discovery/consul/consul.go [hunk 2]
```diff
@@ -25,42 +34,89 @@ func init() {
...
-	p := &consul.KVPair{Key: s.prefix, Value: nil}
-	if _, err = kv.Put(p, nil); err != nil {
-		return err
-	}
-	_, meta, err := kv.Get(s.prefix, nil)
-	if err != nil {
-		return err
-	}
-	s.lastIndex = meta.LastIndex
+	s.prefix = "consul/"
```
Comment (vieux) [inline discovery/consul/consul.go]: why do you hardcode `consul` before we were able to choose it was better. no ?

File: discovery/consul/consul.go (hunk 1 above)
Comment (abronan) [inline discovery/consul/consul.go]: No it's not :D

File: discovery/consul/consul.go (hunk 2 above)
Comment (abronan) [inline discovery/consul/consul.go]: Yes, I will put it back to `path` value.

**Comment by vieux** [comment]
@MrMMorris what do you think of this ? 

Looking at the code, I see it's not supported at all by consul (unlike etcd or zookeeper) so I wonder if we should simply fix the usage and only accept one `addr`

**Comment by vieux** [comment]
same question @david-guenault @JeffChien ^

**Comment by MrMMorris** [comment]
@vieux so consul doesn't allow the multiple IPs?

**Comment by MrMMorris** [comment]
it definitely doesn't make sense to only accept one IP as consul should be clustered

**Comment by vieux** [comment]
@MrMMorris to be confirmed by @abronan but apparently your are supposed to give one IP and then use `chit-chat protocol` to find the others

**Comment by MrMMorris** [comment]
@vieux but then what happens when the initial node you reference disappears?

**Comment by abronan** [comment]
To my limited knowledge, consul client only takes the IP of a single consul server (unlike etcd which takes a list of IPs to create the client). I don't know if on node failure Consul knows about the other agents and switches automatically to another agent. I probably missed something but for now either we create one client per consul agent (painful if you have a lot of them) or we switch to another consul agent by creating a new client on failure.

**Comment by MrMMorris** [comment]
Consul client config allows for specifying multiple Consul server IP's. This is my current client conf:

```json
{
  "start_join": [
    "10.0.0.5",
    "10.0.0.6",
    "10.0.1.5"
],
  "disable_remote_exec": true,
  "domain": "consul.",
  "data_dir": "/opt/consul/data",
  "log_level": "INFO",
  "node_name": "comcast-dev",
  "server": false,
  "client_addr": "0.0.0.0",
  "bind_addr": "0.0.0.0",
  "datacenter": "ec2",
  "rejoin_after_leave": true,
  "leave_on_terminate": false
}
```

**Comment by abronan** [comment]
@MrMMorris Which client are you using? For the Go client, if I take a look at https://github.com/hashicorp/consul/blob/master/api/api.go and the `Config` struct there is only a single string field for the consul server address.

**Comment by MrMMorris** [comment]
@abronan not sure. I am downloading from https://dl.bintray.com/mitchellh/consul/0.5.0_linux_amd64.zip

and setting config based on https://www.consul.io/docs/agent/options.html#start_join

**Comment by abronan** [comment]
From the issue hashicorp/consul#849, Consul go client only takes a single consul server's IP thus we need to handle the failover logic ourselves.

ping @vieux @aluzzardi thoughts?

**Comment by vieux** [comment]
it's your call if you are confident with merging this let's go otherwise we only support one IP

**Comment by abronan** [comment]
I'm a bit skeptical about this and I wouldn't merge it as it is. The proper way to use the consul client is clearly described hashicorp/consul#849 using Consul DNS. This PR is mostly a work-around to do the failover ourselves.

**Comment by abronan** [comment]
@MrMMorris What do you think about this? I'd like first to hear about a good use case to avoid using Consul DNS and just point to a set of Consul servers.

**Comment by MrMMorris** [comment]
@abronan I am fine with having to use Consul DNS as I keep finding more reasons to implement it anyways.

**Comment by abronan** [comment]
@MrMMorris OK thanks for the quick followup. I'm closing the PR for now.

**Pull request closed without merging**